### PR TITLE
rby: Has some interesting optimisations but could be improved further with a custom hash map

### PR DESCRIPTION
**Pull request by rby**
- segments the file to number of cores partitions 
- uses ints instead of doubles
- dumb state machine for parsing
- TODO: a custom hash map could improve the updating operation
- ✅ passes ./test.sh rby

### CODE REVIEW

File: src/main/java/dev/morling/onebrc/CalculateAverage_rby.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+                System.out.print(", " + e.getKey() + "="
+                        + (stats[ix]/10.0) + "/"
+                        + avg + "/"
+                        + (stats[ix + 1]/10.0)) ;
+            }
+            System.out.println("}");
+        }
+    }
+
+    static final int MAX_CITIES = 1000;
```
Comment (gunnarmorling) [inline src/main/java/dev/morling/onebrc/CalculateAverage_rby.java:133]: Hum hum, there can be 10,000 station names. How comes that measurements-10000-unique-keys.txt is passing for this one? CC @AlexanderYastrebov

File: src/main/java/dev/morling/onebrc/CalculateAverage_rby.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+                            strbuffIx = 0;
+                            switch (nextChar) {
+                                case ';':
+                                    nextChar = '\n';
+                                    var mbCityIx = cityIndexes.get(str);
+                                    if (mbCityIx == null) {
+                                        cityIndex = nextCityIx;
+                                        cityIndexes.put(str, nextCityIx++);
+                                        if (nextCityIx * 4 >= stats.length) {
+                                            var newStats = Arrays.copyOf(stats, stats.length * 2);
```
Comment (rby) [inline src/main/java/dev/morling/onebrc/CalculateAverage_rby.java:185]: @gunnarmorling This is where more than 1000 cities is handled.
the array is expanded and copied.
In the main case of 1b with real cities names the number was less than 1000

File: src/main/java/dev/morling/onebrc/CalculateAverage_rby.java (hunk 1 above)
Comment (rby) [reply src/main/java/dev/morling/onebrc/CalculateAverage_rby.java:133]: Hi @gunnarmorling, thanks for checking this. In line 185, the array is expanded if there's an overflow.

Comment (gunnarmorling) [reply src/main/java/dev/morling/onebrc/CalculateAverage_rby.java:133]: Ah yes, got it. Thanks!

**Comment by gunnarmorling** [comment]
36.518sec. Thx for participating!

**Pull request merged**